### PR TITLE
OpenCL: cleanup and rev support for add_diag, unary +-, mean, sub_col, sub_row, segment

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

This PR does dome minor cleanup of /opencl and adds support for a few simple functions. In detail:
- removes copy_triangular, diagonal_multiply, identity and sub_block. The first was not used, other were replaced by k.g.
- added rev support for add_diag, unary operator+ (and plus()), unary operator- (and minus()), mean, sub_col, sub_row, segment

## Tests

All new functions are tested

## Side Effects

/

## Release notes

OpenCL: added rev support for add_diag, unary operator+ (and plus()), unary operator- (and minus()), mean, sub_col, sub_row and segment.

## Checklist

- [x] Math issue #1854 

- [x] Copyright holder: Rok Češnovar

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
Ready for review.

File: stan/math/opencl/tri_inverse.hpp [hunk 1]
```diff
@@ -91,7 +90,8 @@ inline matrix_cl<T> tri_inverse(const matrix_cl<T>& A) {
   // the number of blocks in the first step
   // each block is inverted with using the regular forward substitution
   int parts = inv_padded.rows() / thread_block_size_1D;
-  inv_padded.sub_block(inv_mat, 0, 0, 0, 0, inv_mat.rows(), inv_mat.rows());
+  block_zero_based(inv_padded, 0, 0, inv_mat.rows(), inv_mat.rows())
+      = block_zero_based(inv_mat, 0, 0, inv_mat.rows(), inv_mat.rows());
```
Comment (t4c1) [inline stan/math/opencl/tri_inverse.hpp:94]:
```suggestion
      = inv_mat;
```

File: stan/math/opencl/tri_inverse.hpp [hunk 2]
```diff
@@ -105,11 +105,13 @@ inline matrix_cl<T> tri_inverse(const matrix_cl<T>& A) {
     check_opencl_error("inverse step1", e);
   }
   // set the padded part of the matrix and the upper triangular to zeros
-  inv_padded.sub_block(zero_mat, 0, 0, inv_mat.rows(), 0, zero_mat.rows(),
-                       zero_mat.cols());
+  block_zero_based(inv_padded, inv_mat.rows(), 0, zero_mat.rows(),
+                   zero_mat.cols())
+      = block_zero_based(zero_mat, 0, 0, zero_mat.rows(), zero_mat.cols());
```
Comment (t4c1) [inline stan/math/opencl/tri_inverse.hpp:110]:
```suggestion
      = zero_mat;
```

File: stan/math/opencl/tri_inverse.hpp [hunk 3]
```diff
@@ -105,11 +105,13 @@ inline matrix_cl<T> tri_inverse(const matrix_cl<T>& A) {
...
   inv_padded.template zeros_strict_tri<stan::math::matrix_cl_view::Upper>();
   if (parts == 1) {
-    inv_mat.sub_block(inv_padded, 0, 0, 0, 0, inv_mat.rows(), inv_mat.rows());
+    block_zero_based(inv_mat, 0, 0, inv_mat.rows(), inv_mat.rows())
```
Comment (t4c1) [inline stan/math/opencl/tri_inverse.hpp:113]:
```suggestion
    inv_mat
```

File: stan/math/opencl/tri_inverse.hpp [hunk 4]
```diff
@@ -146,12 +148,14 @@ inline matrix_cl<T> tri_inverse(const matrix_cl<T>& A) {
     }
     result_matrix_dim *= 2;
     // set the padded part and upper diagonal to zeros
-    inv_padded.sub_block(zero_mat, 0, 0, inv_mat.rows(), 0, zero_mat.rows(),
-                         zero_mat.cols());
+    block_zero_based(inv_padded, inv_mat.rows(), 0, zero_mat.rows(),
+                     zero_mat.cols())
+        = block_zero_based(zero_mat, 0, 0, zero_mat.rows(), zero_mat.cols());
```
Comment (t4c1) [inline stan/math/opencl/tri_inverse.hpp:153]:
```suggestion
        = zero_mat;
```

File: stan/math/opencl/tri_inverse.hpp [hunk 5]
```diff
@@ -146,12 +148,14 @@ inline matrix_cl<T> tri_inverse(const matrix_cl<T>& A) {
...
     inv_padded.template zeros_strict_tri<stan::math::matrix_cl_view::Upper>();
   }
   // un-pad and return
-  inv_mat.sub_block(inv_padded, 0, 0, 0, 0, inv_mat.rows(), inv_mat.rows());
+  block_zero_based(inv_mat, 0, 0, inv_mat.rows(), inv_mat.rows())
```
Comment (t4c1) [inline stan/math/opencl/tri_inverse.hpp:157]:
```suggestion
  inv_mat
```

**Review by t4c1 — CHANGES_REQUESTED** [review]

**Review by t4c1 — APPROVED** [review]

**Pull request merged**